font/src/ft/mod.rs

```diff
@@ -529,6 +581,84 @@ impl FreeTypeRasterizer {
     }
 }
 
+fn downsample_bitmap(mut bitmap_glyph: RasterizedGlyph, fixup_factor: f64) -> RasterizedGlyph {
+    // Don't try to upscale
+    if fixup_factor >= 1.0 {
+        return bitmap_glyph;
+    }
+
+    let bitmap_buffer = if let BitmapBuffer::RGBA(bitmap_buffer) = bitmap_glyph.buf {
+        bitmap_buffer
+    } else {
+        return bitmap_glyph;
+    };
+
+    let bitmap_width = bitmap_glyph.width as usize;
+    let bitmap_height = bitmap_glyph.height as usize;
+
+    let width = (bitmap_width as f64 * fixup_factor) as usize;
+    let height = (bitmap_height as f64 * fixup_factor) as usize;
+
+    let advance_step = (1.0 / fixup_factor).ceil() as usize;
+    let mut scaled_buffer = Vec::<u8>::with_capacity(width * height * 4);
+
+    let mut source_line_index = 0;
+
+    for _ in 0..height {
+        let mut source_column_index = 0;
+
+        for _ in 0..width {
+            let mut r: u32 = 0;
+            let mut g: u32 = 0;
+            let mut b: u32 = 0;
+            let mut a: u32 = 0;
+            let mut pixels_picked: u32 = 0;
+
+            let source_end_line = min(source_line_index + advance_step, bitmap_height);
+            let source_end_column = min(source_column_index + advance_step, bitmap_width);
+
+            for source_j in source_line_index..source_end_line {
+                let cur_pixel_index = source_j * bitmap_width * 4;
```

Comment:
I think `source_pixel_index` says much more about this variable than `cur_pixel_index`.